extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/CompletableFutureTest.java

```diff
@@ -0,0 +1,154 @@
+package io.quarkus.smallrye.graphql.deployment;
+
+import static io.quarkus.smallrye.graphql.deployment.AbstractGraphQLTest.MEDIATYPE_JSON;
+import static io.quarkus.smallrye.graphql.deployment.AbstractGraphQLTest.getPropertyAsString;
+
+import java.time.LocalDate;
+import java.time.Month;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.microprofile.graphql.GraphQLApi;
+import org.eclipse.microprofile.graphql.Query;
+import org.eclipse.microprofile.graphql.Source;
+import org.hamcrest.CoreMatchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.restassured.filter.log.ResponseLoggingFilter;
+
+/**
+ * Make sure that CompletableFuture works
+ */
+public class CompletableFutureTest extends AbstractGraphQLTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(BookGraphQLApi.class, Book.class)
+                    .addAsResource(new StringAsset(getPropertyAsString()), "application.properties")
+                    .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml"));
+
+    @Test
+    public void testSourcePost() {
+        String booksRequest = getPayload("{\n" +
+                "    books {\n" +
+                "       title\n" +
+                "       published\n" +
+                "       buyLink\n" +
+                "       authors {\n" +
+                "           name\n" +
+                "           bornName\n" +
+                "       }\n" +
+                "    }\n" +
+                "}");
+
+        RestAssured.given()
+                .filter(new ResponseLoggingFilter())
+                .when()
+                .accept(MEDIATYPE_JSON)
+                .contentType(MEDIATYPE_JSON)
+                .body(booksRequest)
+                .post("/graphql")
+                .then()
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body(CoreMatchers.containsString("Lord of the Flies"),
+                        CoreMatchers.containsString("https://www.amazon.com/s?k=Lord+of+the+Flies&i=stripbooks-intl-ship"), // Test source
+                        CoreMatchers.containsString("William Gerald Golding")); // Test batch
+
+    }
+
+    @GraphQLApi
+    public static class BookGraphQLApi {
+
+        @Query
+        public CompletableFuture<List<Book>> getBooks() {
```

Comment:
The user should not consume CF, but CS. 
CFs are writable, which could lead to interesting ISE.

Reply:
Done